Handle SSM session disconnection w/ AWS get connection status.

Glyphack opened the pull request:
With the current implementation was failing in case of a restart in execution commands

## Solution

With the current implementation was failing in case of a restart in execution commands, for example:
```hcl
build {
  sources = ["source.amazon-ebs.test"]
  provisioner "shell" {
    expect_disconnect = true
    script            = "./reboot.sh"
  }
  provisioner "shell" {
    inline = ["echo 'Hello World!'", "echo >>/home/ec2-user/test-inline.txt"]
  }
}
```
Would not continue after the first script is ran.

To fix this I have changed the code to monitor the instance connection status to SSM service, and in case the session is disconnected it will re-create another session to SMS. This session will be kept open(in case of future crashes) until the packer actually stops the instance.

A few important details:

**How do we identify when to create new session?**

After we run the command to create a [session with port forwarding](https://github.dev/hashicorp/packer-plugin-amazon/blob/3a8a7dd29b0a2ed5bee6636d15c7bb2181098bab/builder/common/ssm/session.go#L111) when the instance restarts then this command will output the error message:
```
amazon-ebs.test: Connection to destination port failed, check SSM Agent logs.
```

After ~1.5 minutes the session manager plugin exits with the message:

```
document process failed unexpectedly: document worker timed out , check [ssm-document-worker]/[ssm-session-worker] log for crash reason
```

After this message we start another SSM session, and the Provision step reconnects to that session, since session is on the same port the reconnection succeeds.

## New Behavior

SSM session will reconnect in case the connection drops until the builder stops the instance(The actual build finishes). In order for the SSM session without a keypair, the ssh public key has to be sent periodically.

## Test

This is the build spec I used for the test:
```
build {
  sources = ["source.amazon-ebs.test"]
  provisioner "shell" {
    expect_disconnect = true
    script            = "./reboot.sh"
  }
  provisioner "shell" {
    inline = ["echo 'Hello World!'", "echo >>/home/ec2-user/test-inline.txt"]
  }
}
```

and reboot script is:
```
echo "Hello I'm rebooting you"

sudo reboot
```

lbajolet-hashicorp commented on builder/common/step_create_ssm_tunnel.go:
```diff
@@ -72,31 +73,61 @@ func (s *StepCreateSSMTunnel) Run(ctx context.Context, state multistep.StateBag)
-		if err != nil {
-			ui.Error(fmt.Sprintf("ssm error: %s", err))
+		for ctx.Err() == nil {
+			ssmAvailable, err := ssmConn.GetConnectionStatus(&ssm.GetConnectionStatusInput{
```
I wonder if we can limit the number of attempts here, from what I understand this will loop without any delay to check the status of the connection.
This will be as responsive as can be from the looks of it, but I think we can probably add some `time.Sleep` to reduce the number of checks we do during a build.

I might be overthinking it though, if you have another idea let me know, always eager to hear suggestions.

Glyphack replied:
This is completely valid. I will try it out with different sleep times and see what works with the least number of calls.

lbajolet-hashicorp reviewed (changes requested):
Hi @Glyphack,

I've taken a look at the code and tested it locally, it does seem to fix the problem of disconnection on a build that uses SSM, thanks for the fix!

I left a few comments on the uploading of the key, I think we can do better than uploading it repeatedly during the build, and only re-upload the key when we have a new SSM session active by adding a channel for communicating between the two coroutines that handle resuming the connection.

I'll let you cycle on this, if you need a hand adapting the code, let me know and I'll take a jab at it.

Thanks for the PR!

Glyphack commented on builder/common/step_create_ssm_tunnel.go:
```diff
@@ -72,31 +73,61 @@ func (s *StepCreateSSMTunnel) Run(ctx context.Context, state multistep.StateBag)
-		if err != nil {
-			ui.Error(err.Error())
-			return multistep.ActionHalt
+	// SSH public key sent expires every minute.
```
Now that I'm thinking, if we only send the key when the connection is open using a channel. Then we can remove this part and the ssm session connector can close the channel when there will be no more connections.

Glyphack commented:
Thanks @lbajolet-hashicorp for the review!

I applied your comment and while doing that I realized I made a mistake too.
The [Start](https://github.com/Glyphack/packer-plugin-amazon/blob/7fa63a9e92f01b33be06665515de3811b729f797/builder/common/ssm/session.go#L92) func previously was wrapping the code in a for loop so the connection was always being recreated, and the code I added was conflicting with this(it was working but it was accidental).

So I made a slight change, I split the function in create session step and moved it as a function to `session.go` this file now has to methods:

- `Start`: to start a session same as before
- `StartWithReconnet`: Which uses the `Start` but reconnects. The status check is every 30 seconds which worked fine for my machine, it starts very shortly after the reboot is done.

The channel is also created for the communication as you suggested. I used the `struct{}` as type since it's not sending any info, what do you think?

lbajolet-hashicorp reviewed (changes requested):
Hi @Glyphack,

I've been experimenting with your branch and writing some more acceptance tests, this gives me some more perspective on the changes you are proposing.

From what I see, re-uploading the public key derived from the private one defined in the template works when re-establishing the SSM session, I have not yet been able to reproduce a disconnection that does not trigger reinitialising the SSM session.

Regarding the re-establishement of an SSM connection when we lose the connection because of a reboot for example, the system recovers on its own, albeit maybe after a while, depending on how long we wait for the VM to reboot and the ssm-agent to start.
Once both are up, the current code re-establishes the session, and is able to continue with the build (except in the case where the SSH private key was read from a file).

All in all, I think your PR should focus on changing the logic for triggering the publication of the SSH key, so that it is done either periodically (this might be more resilient, but my testing failed to highlight a problem so take that with a grain of salt), or when the connection is re-established.

I would suggest we leave `ssm.Session.Start` as-is  for now, as you pointed out, it already re-establishes the connection here, but I do think we can improve it by adding the channel that we are listening on to be aware that we should re-upload the SSH key when it is re-established.

This or maybe go back with the `Ticker` and a loop/select so we push keys periodically until the context gets Done.

I'll push the acceptance tests I wrote for this feature on top of your branch so you can take a look at them for reference.
The `TestAccBuilder_EbsSSMRebootProvisioner` test succeeds on both `main` and your branch, while `TestAccBuilder_PrivateKeyFileWithReboot` does not since the key isn't re-uploaded.

Again thank you for the reroll here, let me know if you want me to work on this code.

Glyphack commented on builder/common/ssm/session.go:
```diff
@@ -90,28 +90,53 @@ func (s Session) getCommand(ctx context.Context) ([]string, string, error) {
+				ui.Error(fmt.Sprintf("ssm error: %s", err))
+			}
 		}
+		time.Sleep(30 * time.Second)
```
Applied

Glyphack commented on builder/common/ssm/session.go:
```diff
@@ -90,28 +90,53 @@ func (s Session) getCommand(ctx context.Context) ([]string, string, error) {
+		case "connected":
+			ui.Say(fmt.Sprintf("ssm: start PortForwarding session to instance %s", s.InstanceID))
+			sessionChan <- struct{}{}
+			err = s.Start(ctx, ui)
```
No you are right, this loop was useful just before we start the session, basically before the first session it was checking periodicly if the session manager has connected to instance or not.

I removed this part completely as part of your comment about not touching the Start function in this PR

Glyphack commented:
Hi @lbajolet-hashicorp

I have reverted the parts that I modified in the `ssm.go` as you suggested, the PR was getting really big and I could have introduced side effects that was not easy to fix 😄.

Now the changes ensure that the key is sent upon reconnects the `Start` function only has one change that now it takes in a channel for sending create session signals.
Also since this channel must be always emptied otherwise it blocks the session goroutine, the checks for sending the public key [are done](https://github.com/hashicorp/packer-plugin-amazon/pull/311/commits/f07a5b27986f0d6c1505bdb224c65f424f8d9d8c) inside the loop.

I tested it with and without keys and it runs fine.

Thanks for the guidance. let me know your thoughts.

lbajolet-hashicorp reviewed (approved):
Hi @Glyphack,

Thanks for the speedy reroll, focusing on the re-emission of the key looks good to me.

I was about to ask if you wanted my help rebasing these changes on top of main, but I see that in the meantime you fixed it all, so we should be good to go with merging this PR.

Thanks again for the contribution, since this was re-focused around uploading SSH keys, this will probably not fix the linked issue, so I will remove the mention in the PR message before merging this, I'll follow-up on the issue so we have a template to reproduce the error with and we can then focus on fixing this problem.

lbajolet-hashicorp commented:
FYI @Glyphack, I pushed a slight change with the code since lint checks complained that the error for `session.Start` wasn't checked.

Acceptance tests run as expected, and the tests are now passing, I'm merging this PR now
